libs/full/async_cuda/src/cuda_event_callback.cpp

```diff
@@ -0,0 +1,257 @@
+//  Copyright (c) 2021 ETH Zurich
+//  Copyright (c) 2020 John Biddiscombe
+//  Copyright (c) 2016 Hartmut Kaiser
+//  Copyright (c) 2016 Thomas Heller
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/config.hpp>
+#include <hpx/assert.hpp>
+#include <hpx/async_cuda/cuda_event.hpp>
+#include <hpx/async_cuda/custom_gpu_api.hpp>
+#include <hpx/async_cuda/detail/cuda_debug.hpp>
+#include <hpx/async_cuda/detail/cuda_event_callback.hpp>
+#include <hpx/concurrency/concurrentqueue.hpp>
+#include <hpx/synchronization/spinlock.hpp>
+#include <hpx/threading_base/scheduler_base.hpp>
+#include <hpx/threading_base/thread_pool_base.hpp>
+
+#include <algorithm>
+#include <atomic>
+#include <cstddef>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace hpx { namespace cuda { namespace experimental { namespace detail {
+    // this code runs on a std::thread, but we will use a spinlock
+    // as we never suspend - only ever try_lock, or exit
+    using mutex_type = hpx::lcos::local::spinlock;
+    mutex_type& get_vector_mtx()
+    {
+        static mutex_type vector_mtx;
+        return vector_mtx;
+    }
+
+#if defined(HPX_DEBUG)
+    std::atomic<std::size_t>& get_register_polling_count()
+    {
+        static std::atomic<std::size_t> register_polling_count{0};
+        return register_polling_count;
+    }
+#endif
+
+    // Holds a CUDA event and a callback. The callback is intended to be called when
+    // the event is ready.
+    struct event_callback
+    {
+        cudaEvent_t event;
+        event_callback_function_type f;
+    };
+
+    using event_callback_queue_type =
+        concurrency::ConcurrentQueue<event_callback>;
+    using event_callback_vector_type = std::vector<event_callback>;
+
+    event_callback_vector_type& get_event_callback_vector()
+    {
+        static event_callback_vector_type event_callback_vector;
+        return event_callback_vector;
+    }
+
+    event_callback_queue_type& get_event_callback_queue()
+    {
+        static event_callback_queue_type event_callback_queue;
+        return event_callback_queue;
+    }
+
+    void add_to_event_callback_queue(event_callback&& continuation)
+    {
+        HPX_ASSERT_MSG(get_register_polling_count() != 0,
+            "CUDA event polling has not been enabled on any pool. Make sure "
+            "that CUDA event polling is enabled on at least one thread pool.");
+
+        get_event_callback_queue().enqueue(std::move(continuation));
+
+        cud_debug.debug(debug::str<>("event queued"), "event",
+            debug::hex<8>(continuation.event), "callbacks in vector",
```

Comment:
callbacks in queue?

Reply:
It's using `get_event_callack_vector`, so seems ok to me?

Reply:
Okay I thought you would want to print the size of the queue here as it's the queue that is modified but I might be mistaken

Reply:
I think you're right, it looks like it would make more sense to print the size of the queue!